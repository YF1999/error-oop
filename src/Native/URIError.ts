import { URIErrorProps } from '../CommonTypes';
import { appendInnerErrorStack, setNonEnumerable } from './Tools';

export abstract class AbstractURIError extends URIError {
    #innerError?: Error;

    public constructor(props: URIErrorProps) {
        super(props.message);

        this.#innerError = props.innerError;

        // When the first call to `stack` property happens, it will combine `name` and `message` with trace stack to
        // `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this.#innerError);

        this._setNonEnumerable('stack');
    }

    public get name(): string {
        return this.constructor.name;
    }

    public get innerError(): Error | undefined {
        return this.#innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

/**
 * Represents an error when a value is not of the expected type. This is roughly the same as the native URIError class.
 * It additionally supports an innerError attribute.
 */
export class URIErrorPro extends AbstractURIError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, innerError: Error);

    public constructor(message: string = '', innerError?: Error) {
        super({ message, innerError });
    }
}

export { URIErrorPro as URIError };
