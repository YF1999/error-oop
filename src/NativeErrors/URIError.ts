import { ErrorOptions, URIErrorMessageProps, URIErrorProps } from '../CommonTypes';
import { appendInnerErrorStack, setNonEnumerable } from '../utils';

export abstract class AbstractURIError extends URIError {
    protected _innerError?: Error;

    public constructor(props: URIErrorProps, options: ErrorOptions<URIErrorMessageProps>) {
        super();

        const { message, innerError } = props;
        const { name, generateMessage } = options;

        this._innerError = innerError;
        this.name = name || this.constructor.name;
        this.message = generateMessage ? generateMessage({ name: this.name, message }) : message;

        // When the first call to `stack` property happens, it will combine `name` and `message` with trace stack to
        // `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this._innerError);

        this._setNonEnumerable('name');
        this._setNonEnumerable('message');
        this._setNonEnumerable('stack');
        this._setNonEnumerable('_innerError');
    }

    public get innerError(): Error | undefined {
        return this._innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

/**
 * Represents an error when a value is not of the expected type. This is roughly the same as the native URIError class.
 * It additionally supports an innerError attribute.
 */
export class URIErrorPro extends AbstractURIError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, innerError: Error);

    public constructor(message: string = '', innerError?: Error) {
        super({ message, innerError }, { name: 'URIError' });
    }
}

export { URIErrorPro as URIError };
