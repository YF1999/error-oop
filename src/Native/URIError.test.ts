import { NativeURIError } from './URIError';

standardTestError(NativeURIError, 'NativeURIError', 'error');
