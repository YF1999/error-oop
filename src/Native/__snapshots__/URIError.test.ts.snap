// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`NativeURIError\` own keys 1`] = `
Array [
  "stack",
]
`;

exports[`Test \`NativeURIError\` own keys 2`] = `
Array [
  "stack",
  "message",
]
`;

exports[`Test \`NativeURIError\` own keys 3`] = `
Array [
  "stack",
  "message",
]
`;

exports[`Test \`NativeURIError\` own keys 4`] = `
Array [
  "stack",
  "message",
]
`;
